src/server/model.coffee

```diff
@@ -502,26 +502,35 @@ module.exports = Model = (db, options) ->
         return callback? error if error
 
         docTemplate = {v: 0, type: type, snapshot: "", meta: null}
-        results = []
-
-        try
-          for op in ops
-            docTemplate.v = op.v + 1
-            docTemplate.snapshot = type.apply(docTemplate.snapshot, op.op)
-            docTemplate.meta = op.meta
-
-            if docTemplate.v % n is 0
-              results.push({
-                v: docTemplate.v,
-                type: docTemplate.type.name,
-                snapshot: docTemplate.snapshot,
-                meta: docTemplate.meta
-              })
-
-          callback? null, results
-        catch e
-          console.error "Op data invalid for #{docName}: #{e.stack}"
-          callback? 'Op data invalid'
+        opBatchAmount = 1000
+
+        buildVersions = (callback, ops, results = []) ->
+          if ops.length
+            versionOps = ops.slice(0, opBatchAmount)
+
+            try
+              for op in versionOps
+                docTemplate.v = op.v + 1
+                docTemplate.snapshot = type.apply(docTemplate.snapshot, op.op)
+                docTemplate.meta = op.meta
+
+                if docTemplate.v % n is 0
+                  results.push({
+                    v: docTemplate.v,
+                    type: docTemplate.type.name,
+                    snapshot: docTemplate.snapshot,
+                    meta: docTemplate.meta
+                  })
+
+              setTimeout buildVersions.bind(null, callback, ops.slice(opBatchAmount), results)
+            catch e
+              console.error "Op data invalid for #{docName}: #{e.stack}"
+              callback? 'Op data invalid'
+
+          else
+            callback? null, results
+
+        setTimeout buildVersions.bind(null, callback, ops)
```

Comment:
just to confirm... as this is such a trip seeing it... this just pushing the batch onto the call stack... to be processed async yeah.

Reply:
and also... only pops 1000 off ops at a time?

Reply:
yeah, would have preferred to use a library or async/await but considering how old this stack is I did it the old fashioned way :)

Reply:
yep, that works out to about 100ms per batched "job". Just keeps it from clogging up the event loop.

Again though, really temporary fix to prevent the timeouts and connection problems while we come up with a better solution (which probably involves not trying to rebuild entire documents for each snapshot step).

Reply:
👍